### PR TITLE
Refactor/types test

### PR DESCRIPTION
**Pull request by rimashah25**
This PR refactors v4/types_test.go, v3/types_test.go, in order to make tests more granular, providing transparency on what is being tested and reducing the need to dig through test code to confirm what was tested. In addition this new test structure should also make it easier for new tests to be added as well as reduce redundant code.

- Adds data driven testing
   - Descriptive test names 
- Sub tests for organization
   - Use GoLangs test runner
- Use assert functionality
    - Provides fundamental assertions
    - Streamlines code
    - Reduces nested conditionals
- Adds expectation functions
   - Test specific expectations in a clear concise way
   - Reusable expectations

<hr/>

## Which Traffic Control components are affected by this PR?

- Traffic Ops

## What is the best way to verify this PR?

Run TO Integration Tests

## PR submission checklist
- [x] This PR has tests 
- [ ] This PR has documentation 
- [ ] This PR has a CHANGELOG.md entry 
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

### CODE REVIEW

File: traffic_ops/testing/api/v3/types_test.go [hunk 1]
```diff
@@ -28,224 +34,255 @@ import (
...
 		err := db.Close()
-		if err != nil {
-			t.Errorf("unable to close connection to db, error: %v", err)
-		}
+		assert.NoError(t, err, "unable to close connection to db, error: %v", err)
 	}()
-	dbQueryTemplate := "INSERT INTO type (name, description, use_in_table) VALUES ('%v', '%v', '%v');"
+	dbQueryTemplate := "INSERT INTO type (name, description, use_in_table) VALUES ('%s', '%s', '%s');"
 
+	//opts := client.NewRequestOptions()
```
Comment (ericholguin) [inline traffic_ops/testing/api/v3/types_test.go:243]: Can remove comment

File: traffic_ops/testing/api/v3/types_test.go [hunk 2]
```diff
@@ -28,224 +34,255 @@ import (
...
 
-		// Retrieve the Type to see if it got deleted
-		types, _, err := TOSession.GetTypeByName(typ.Name)
-		if err != nil {
-			t.Errorf("error deleting Type name: %v", err)
-		}
-		if len(types) > 0 {
-			t.Errorf("expected Type name: %s to be deleted", typ.Name)
-		}
+		// Retrieve the Type by name so we can get the id for the Update
```
Comment (ericholguin) [inline traffic_ops/testing/api/v3/types_test.go:281]: Comment should be updated to: // Retrieve the Type to see if it got deleted

File: traffic_ops/testing/api/v4/types_test.go [hunk 3]
```diff
@@ -16,243 +16,264 @@ package v4
...
-			delResp, _, err := TOSession.DeleteType(respType.ID, client.RequestOptions{})
-			if err != nil {
-				t.Fatalf("cannot delete Type using the API: %v - alerts: %+v", err, delResp.Alerts)
-			}
+			delResp, _, err := TOSession.DeleteType(typ.ID, client.RequestOptions{})
+			assert.RequireNoError(t, err, "cannot delete Type using the API: %v - alerts: %+v", err, delResp.Alerts)
 		}
 
-		// Retrieve the Type to see if it got deleted
+		// Retrieve the Type by name so we can get the id for the Update
```
Comment (ericholguin) [inline traffic_ops/testing/api/v4/types_test.go:272]: Comment should be updated to: // Retrieve the Type to see if it got deleted

File: traffic_ops/testing/api/v4/types_test.go [hunk 4]
```diff
@@ -16,243 +16,264 @@ package v4
...
 		err := db.Close()
-		if err != nil {
-			t.Errorf("unable to close connection to db, error: %v", err)
-		}
+		assert.NoError(t, err, "unable to close connection to db, error: %v", err)
 	}()
 	dbQueryTemplate := "INSERT INTO type (name, description, use_in_table) VALUES ('%s', '%s', '%s');"
 
-	opts := client.NewRequestOptions()
+	//opts := client.NewRequestOptions()
```
Comment (ericholguin) [inline traffic_ops/testing/api/v4/types_test.go:234]: Can remove comment

File: traffic_ops/testing/api/v3/types_test.go [hunk 5]
```diff
@@ -28,224 +34,255 @@ import (
...
+						"lastUpdated": "2022-06-17T19:13:46.802044+00:00",
+						"name":        "TEST_4",
+						"useInTable":  "server",
+					},
+					Expectations: utils.CkRequest(utils.NoError(), utils.HasStatus(http.StatusOK),
+						validateTypeUpdateCreateFields("TEST_4", map[string]interface{}{"Name": "TEST_4"})),
+				},
+			},
+			"PUT": {
+				"BAD REQUEST when NOT in server table": {
```
Comment (ericholguin) [inline traffic_ops/testing/api/v3/types_test.go:84]: nit: Can we change this to: "BAD REQUEST when USEINTABLE NOT server". Same for POST and v4

File: traffic_ops/testing/api/v3/types_test.go [hunk 6]
```diff
@@ -28,224 +34,255 @@ import (
...
+					EndpointId:    GetTypeID(t, "ACTIVE_DIRECTORY"),
+					ClientSession: TOSession,
+					RequestBody: map[string]interface{}{
+						"description": "Active Directory User",
+						"name":        "TEST_3",
+						"useInTable":  "cachegroup",
+					},
+					Expectations: utils.CkRequest(utils.HasError(), utils.HasStatus(http.StatusBadRequest)),
+				},
+				"OK when VALID request in SERVER table": {
```
Comment (ericholguin) [inline traffic_ops/testing/api/v3/types_test.go:94]: nit: Can we change this to: "OK when VALID request when USEINTABLE=server". Same for POST and v4

**Review by ericholguin — APPROVED** [review]
LGTM! Can be merged once tests finish passing.

**Review by ocket8888 — APPROVED** [review]

**Pull request merged**